sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java

```diff
@@ -0,0 +1,248 @@
+   */
+  private SQLExecEnvironment(DAG dag)
+  {
+    this.dag = dag;
+    this.typeFactory = new JavaTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+  }
+
+  /**
+   * Given SQLExec{@link SQLExecEnvironment} object for given {@link DAG}.
+   *
+   * @param dag Object of type {@link DAG}
+   * @return Returns {@link SQLExecEnvironment} object
+   */
+  public static SQLExecEnvironment getEnvironment(DAG dag)
+  {
+    return new SQLExecEnvironment(dag);
+  }
+
+  /**
+   * Use given model file to initialize {@link SQLExecEnvironment}.
+   * The model file contains definitions of endpoints and data formats.
+   * Example of file format is like following:
+   * <pre>
+   *   {
+   *     version: '1.0',
+   *     defaultSchema: 'APEX',
+   *     schemas: [{
+   *       name: 'APEX',
+   *       tables: [
+   *         {
+   *            name: 'ORDERS',
+   *            type: 'custom',
+   *            factory: 'org.apache.apex.malhar.sql.schema.ApexSQLTableFactory',
+   *            stream: {
+   *              stream: true
+   *            },
+   *            operand: {
+   *              endpoint: 'file',
+   *              messageFormat: 'csv',
+   *              endpointOperands: {
+   *                directory: '/tmp/input'
+   *              },
+   *              messageFormatOperands: {
+   *                schema: '{"separator":",","quoteChar":"\\"","fields":[{"name":"RowTime","type":"Date","constraints":{"format":"dd/MM/yyyy hh:mm:ss"}},{"name":"id","type":"Integer"},{"name":"Product","type":"String"},{"name":"units","type":"Integer"}]}'
+   *            }
+   *            }
+   *         }
+   *       ]
+   *     }]
+   *   }
+   * </pre>
+   *
+   * @param model String content of model file.
+   * @return Returns this {@link SQLExecEnvironment}
+   */
+  public SQLExecEnvironment withModel(String model)
+  {
+    if (model == null) {
+      return this;
+    }
+    Connection connection = null;
+    try {
+      Properties p = new Properties();
+      p.put("model", "inline:" + model);
+      connection = DriverManager.getConnection("jdbc:calcite:", p);
+      CalciteConnection conn = connection.unwrap(CalciteConnection.class);
+      this.schema = conn.getRootSchema().getSubSchema(connection.getSchema());
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      try {
+        connection.close();
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+    }
+
+    return this;
+  }
+
+  /**
+   * Register a table using {@link Endpoint} with this {@link SQLExecEnvironment}
+   *
+   * @param name Name of the table that needs to be present with SQL Statement.
+   * @param endpoint Object of type {@link Endpoint}
+   *
+   * @return Returns this {@link SQLExecEnvironment}
+   */
+  public SQLExecEnvironment registerTable(String name, Endpoint endpoint)
+  {
+    registerTable(name, new ApexSQLTable(schema, name, endpoint));
```

Comment:
registerTable, registerFunction and withModel are 2 different ways to provide information to Calcite schema. So its absolutely valid to call registerTable without calling withModel.